### ISSUE TITLE
Always reformat tab characters as space characters

### ISSUE THREAD

**Issue by jamesquilty** [executablebooks/mdformat#262]
### Description / Summary

The tab character (0x09) is a pest.

Currently, `mdformat` seeks to "apply consistent white space across the board" ([Formatting Style: Whitespace](https://mdformat.readthedocs.io/en/stable/users/style.html#whitespace)) and does the right thing when tab characters appear as leading white space for indentation: it eliminates the pest by replacing them with the appropriate number of space characters for indentation. Line-trailing tabs are also eliminated.

Unfortunately, tab characters in heading and paragraph bodies, where HTML white space collapse will apply when the HTML is rendered for display, are not eliminated by collapsing them into a single space character. I think that tab characters should be eliminated in this context also because _tabs cause problems_.

I believe there are three contexts where tab characters might appear and there's a case for elimination in each:

| Context | Action |
| --- | --- |
| 1. Line-leading white space for Indentation | Eliminate, replacing with the appropriate number of space characters. This is the current behaviour. |
| 2. Wherever HTML white space collapse applies | Eliminate, replacing with a single space character. This is the proposed enhancement. |
| 3. Wherever HTML white space collapse will not apply, e.g. [Code spans](https://spec.commonmark.org/0.30/#code-spans), [Fenced code blocks](https://spec.commonmark.org/0.30/#fenced-code-blocks) | Either (a) preserve, allowing the HTML renderer to determine how to display tab characters in `<code>` or `<pre>` blocks _or_ (b) expand to the appropriate number of space characters. |

I'd propose that _always_ eliminating tab characters and replacing them with the appropriate number of space characters is the way to "apply consistent white space across the board" and that the current mixed treatment of tab characters is inconsistent with `mdformat`'s style goals. Mixed tabs and spaces are seldom good.

There might be an open question with regard to (3), above, because CSS _might_ change the width of tab characters rendered in `<code>` or `<pre>` or other HTML blocks?

### Value / benefit

* Consistent treatment of white space introduced by tab characters.
* Markdown source will more closely resemble the rendered output.
* Avoids display problems in editors caused by differing tab width settings and different invisible white space.
* Produces output which will not violate Markdownlint rule [MD010 - No Hard tabs](https://github.com/markdownlint/markdownlint/blob/master/docs/RULES.md#md010---hard-tabs)

### Implementation details

I think that modifying the `TextWrapper`  instance attributes here

https://github.com/executablebooks/mdformat/blob/a856f538e2dcb81a83e9013fb073f16cd6e53972/src/mdformat/renderer/_context.py#L330-L336

to

```python
        expand_tabs=True,
        tabsize=1
```

will achieve the desired white space collapse of a tab character to a space character, but it won't help in collapsing multiple tab-and-space character runs into a single space. The `replace_whitespace` instance attribute would seem to affect _all_ white space characters and not just tab characters.

**Comment by jamesquilty** [comment]
The [CommonMark Spec 0.20: Preprocessing](https://spec.commonmark.org/0.20/#preprocessing) used to specify:

> Tabs in lines are immediately expanded to spaces, with a tab stop of 4 characters:

but this was changed in version 2.1 onward. I'm not sure what the motivation was for the change, but there two relevant issues on the CommonMark GitHub Project: [commonmark-spec#386](https://github.com/commonmark/commonmark-spec/issues/386) and [commonmark-spec#318](https://github.com/commonmark/commonmark-spec/issues/318).

It's probably worth noting that any tab characters that ever find their way into my Markdown documents are introduced by copy-and-paste and aren't there intentionally.

**Comment by jgopel** [comment]
Has any progress been made on this? I'm very interested in this feature and I'd be open to making a PR if there's interest.

**Comment by jamesquilty** [comment]
@jgopel, I'm still interested in this feature. I don't think that it's been implemented independently of this Issue.

**Comment by jgopel** [comment]
@hukkin Would you be interested in merging this if I were to make a PR for it?

**Comment by hoijui** [comment]
Expanding tabs to spaces is a task that you can do very easily with a bunch of standard POSIX/UNIX utilities. According to the UNIX philosophy (or say: best practice), that is what you should be doing. I would say, this is even a prime example of where this should be applied.

**Comment by jamesquilty** [comment]
@hoijui No, thank you. Please don't butt-in and tell people what they should be doing, it's inappropriate.

**Comment by hoijui** [comment]
where did I do that?

**Comment by hukkin** [comment]
Hey, yeah this feature is welcome provided we test extensively to make sure rendered output never changes in some obscure corner case. Fenced code blocks should not be touched.

**Comment by hukkin** [comment]
https://github.com/executablebooks/mdformat/pull/466 does the conversion for text inlines. I think this should be done to code spans too, so I'll leave this open.

**Comment by jamesquilty** [comment]
@hukkin, Thanks for accepting this and opening a PR! Just doing (2), above, will be a big step forward.

 I think that it might create odd situation if code spans and fenced code blocks were treated differently. My [comment (above)](https://github.com/hukkin/mdformat/issues/262#issuecomment-913927858) links to [then-current] prior discussion on the topic in the [commonmark-spec](https://github.com/commonmark/commonmark-spec) repository. I'm not sure whether the handling of tabs remains an open question.
